File: project/android/app/src/main/java/com/remobile/phone/RCTPhone.java
package com.remobile.phone;

import android.app.Activity;
import android.content.Context;
import android.media.AudioManager;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.gyyx.edutech.bearIFCallback;
import com.gyyx.edutech.bearInterface;
import com.gyyx.edutech.bearViewInterface;


public class RCTPhone extends ReactContextBaseJavaModule implements bearViewInterface {
    public static final String EVENT_ON_JOIN_ROOM = "EVENT_ON_JOIN_ROOM";
    public static final String EVENT_ON_EXIT_ROOM = "EVENT_ON_EXIT_ROOM";
    public static final String EVENT_ON_CONNECT = "EVENT_ON_CONNECT";
    public static final String EVENT_ON_DISCONNECT = "EVENT_ON_DISCONNECT";
    public static final String EVENT_ON_BROADCAST = "EVENT_ON_BROADCAST";
    public static final String EVENT_ON_SEND_MESSAGE = "EVENT_ON_SEND_MESSAGE";
    public static final String EVENT_ON_MATCHING_ROOM = "EVENT_ON_MATCHING_ROOM";
    public static final String EVENT_ON_SET_GRADE = "EVENT_ON_SET_GRADE";
    public static final String EVENT_ON_REQUEST = "EVENT_ON_REQUEST";

    public static final String FLOG_TAG = "RCTPhone";


    private Activity activity;
    // private bearInterface bear;
    // private bearIFCallback bearCallback;
    private int currVolume;

    public RCTPhone(ReactApplicationContext reactContext, Activity activity) {
        super(reactContext);
        this.activity = activity;
        // bear = new bearInterface();
        // bearCallback = new bearIFCallback(this);
    }

    @Override
    public String getName() {
        return "Phone";
    }

    @ReactMethod
    public void connectAudioServer(ReadableMap args) {
        String serverIP = args.getString("serverIP");
        int serverPort = args.getInt("serverPort");
        // bear.setParam(serverIP, serverPort, activity.getApplicationContext(), bearCallback);
    }

    @ReactMethod
    public int joinChatroom(ReadableMap args) {
        String options = args.getString("options");
        String chatroomID = args.getString("chatroomID");

        // return bear.joinChatroom(options, chatroomID);
        return 0;
    }

    @ReactMethod
    public int exitChatroom(ReadableMap args) {
        String userID = args.getString("userID");
        String chatroomID = args.getString("chatroomID");
        // return bear.exitChatroom(userID, chatroomID);
        return 0;
    }

    @ReactMethod
    public int setMessageToServer(ReadableMap args) {
        String userID = args.getString("userID");
        String chatroomID = args.getString("chatroomID");
        int message = args.getInt("message");

        // return bear.setMessageToServer(userID, chatroomID, message);
        return 0;
    }

    @ReactMethod
    public int matchingChatroom(ReadableMap args) {
        String options = args.getString("options");

        // return bear.matchingChatroom(options);
        return 0;
    }

    @ReactMethod
    public int setGradeToServer(ReadableMap args) {
        String userID = args.getString("userID");
        String chatroomID = args.getString("chatroomID");
        String gradeUserID = args.getString("gradeUserID");
        int gradeCount = args.getInt("gradeCount");

        // return bear.setGradeToServer(userID, chatroomID, gradeUserID, gradeCount);
        return 0;
    }

    @ReactMethod
    public int pausePlayout() {
        // return bear.pausePlayout();
        return 0;
    }

    @ReactMethod
    public int resumePlayout() {
        // return bear.resumePlayout();
        return 0;
    }

    @ReactMethod
    public int pauseRecord() {
        // return bear.pauseRecord();
        return 0;
    }

    @ReactMethod
    public int resumeRecord() {
        // return bear.resumeRecord();
        return 0;
    }

    @ReactMethod
    public int speakerOn() {
        try {
            AudioManager audioManager =	(AudioManager) activity.getSystemService(Context.AUDIO_SERVICE);
            audioManager.setMode(AudioManager.ROUTE_SPEAKER);
            currVolume = audioManager.getStreamVolume(AudioManager.STREAM_VOICE_CALL);
            if(!audioManager.isSpeakerphoneOn()) {
                audioManager.setSpeakerphoneOn(true);
                audioManager.setStreamVolume(AudioManager.STREAM_VOICE_CALL,
                        audioManager.getStreamMaxVolume(AudioManager.STREAM_VOICE_CALL ),
                        AudioManager.STREAM_VOICE_CALL);

            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return 1;
    }

    @ReactMethod
    public int speakerOff() {
        try {
            AudioManager audioManager =	(AudioManager) activity.getSystemService(Context.AUDIO_SERVICE);
            audioManager.setMode(AudioManager.MODE_IN_CALL);
            if(audioManager.isSpeakerphoneOn()) {
                audioManager.setSpeakerphoneOn(false);
                audioManager.setStreamVolume(AudioManager.STREAM_VOICE_CALL,currVolume, AudioManager.STREAM_VOICE_CALL);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return 1;
    }
    @ReactMethod
    public int request(String data) {
        // return bear.request(data);
        return 0;
    }

    public int onJoinChatroomRS(final int result, final String chatroomID, final String userID) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                WritableMap params = Arguments.createMap();
                params.putInt("result", result);
                params.putString("chatroomID", chatroomID);
                params.putString("userID", userID);
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_ON_JOIN_ROOM, params);
            }
        });
        return 1;
    }
    public int onExitChatroomRS(final int result, final String chatroomID, final String userID) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                WritableMap params = Arguments.createMap();
                params.putInt("result", result);
                params.putString("chatroomID", chatroomID);
                params.putString("userID", userID);
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_ON_EXIT_ROOM, params);
            }
        });
        return 1;
    }
    public int onAudioServerConnect() {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_ON_CONNECT, true);
            }
        });
        return 1;
    }
    public int onAudioServerDisconnect() {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_ON_DISCONNECT, true);
            }
        });
        return 1;
    }

    public int onAudioServerBroadcast(final String resultJson) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_ON_BROADCAST, resultJson);
            }
        });
        return 1;
    }
    public int onSendMessageToServerRS(final int result, final int message) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                WritableMap params = Arguments.createMap();
                params.putInt("result", result);
                params.putInt("message", message);
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_ON_SEND_MESSAGE, params);
            }
        });
        return 1;
    }
    public int onMatchingChatroomRS(final int result, final String chatroomID, final String userID) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                WritableMap params = Arguments.createMap();
                params.putInt("result", result);
                params.putString("chatroomID", chatroomID);
                params.putString("userID", userID);
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_ON_MATCHING_ROOM, params);
            }
        });
        return 1;
    }
    public int onSetGradeToServerRS(final int result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                WritableMap params = Arguments.createMap();
                params.putInt("result", result);
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_ON_SET_GRADE, params);
            }
        });
        return 1;
    }
    public int onRequestRS(final int result, final String resultJson)
    {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                WritableMap params = Arguments.createMap();
                params.putInt("result", result);
                params.putString("data", resultJson);
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(EVENT_ON_REQUEST, params);
            }
        });
        return 1;
    }
}

File: project/ios/YXjqd/Plugins/phone/RCTPhone.m
#import <React/RCTBridge.h>
#import <React/RCTUtils.h>
#import <React/RCTEventDispatcher.h>
//#include "bear_interface_ios.h"

#define EVENT_ON_JOIN_ROOM @"EVENT_ON_JOIN_ROOM"
#define EVENT_ON_EXIT_ROOM @"EVENT_ON_EXIT_ROOM"
#define EVENT_ON_CONNECT @"EVENT_ON_CONNECT"
#define EVENT_ON_DISCONNECT @"EVENT_ON_DISCONNECT"
#define EVENT_ON_BROADCAST @"EVENT_ON_BROADCAST"
#define EVENT_ON_SEND_MESSAGE @"EVENT_ON_SEND_MESSAGE"
#define EVENT_ON_MATCHING_ROOM @"EVENT_ON_MATCHING_ROOM"
#define EVENT_ON_SET_GRADE @"EVENT_ON_SET_GRADE"
#define EVENT_ON_REQUEST @"EVENT_ON_REQUEST"


@interface RCTPhone : NSObject <RCTBridgeModule>

@end

@implementation RCTPhone

@synthesize bridge = _bridge;

- (id)init {
  self = [super init];
  if (self) {
  }
  return self;
}

RCT_EXPORT_MODULE(Phone);


RCT_EXPORT_METHOD(connectAudioServer:(NSDictionary *)data) {
  NSString *serverIP = data[@"serverIP"];
  int serverPort = [data[@"serverPort"] intValue];
//  [[BearInterface getInstance] initNetioIos:serverIP :serverPort :self];
}

RCT_EXPORT_METHOD(joinChatroom:(NSDictionary *)data) {
  NSString *options = data[@"options"];
  NSString *chatroomID = data[@"chatroomID"];
//  [[BearInterface getInstance] joinChatroom:options :chatroomID];
}

RCT_EXPORT_METHOD(exitChatroom:(NSDictionary *)data) {
  NSString *userID = data[@"userID"];
  NSString *chatroomID = data[@"chatroomID"];
//  [[BearInterface getInstance] exitChatroom:userID :chatroomID];
}

RCT_EXPORT_METHOD(setMessageToServer:(NSDictionary *)data) {
  NSString *userID = data[@"userID"];
  NSString *chatroomID = data[@"chatroomID"];
  int message = [data[@"message"] intValue];
//  [[BearInterface getInstance] setMessageToServer:userID :chatroomID :message];
}

RCT_EXPORT_METHOD(matchingChatroom:(NSDictionary *)data) {
  NSString *options = data[@"options"];
//  [[BearInterface getInstance] matchingChatroom:options];
}

RCT_EXPORT_METHOD(setGradeToServer:(NSDictionary *)data) {
  NSString *userID = data[@"userID"];
  NSString *chatroomID = data[@"chatroomID"];
  NSString *gradeUserID = data[@"gradeUserID"];
  int gradeCount = [data[@"gradeCount"] intValue];
//  [[BearInterface getInstance] setGradeToServer:userID :chatroomID :gradeUserID :gradeCount];
}

RCT_EXPORT_METHOD(pausePlayout) {
//  [[BearInterface getInstance] pausePlayout];
}

RCT_EXPORT_METHOD(resumePlayout) {
//  [[BearInterface getInstance] resumePlayout];
}

RCT_EXPORT_METHOD(pauseRecord) {
//  [[BearInterface getInstance] pauseRecord];
}

RCT_EXPORT_METHOD(resumeRecord) {
//  [[BearInterface getInstance] resumeRecord];
}


RCT_EXPORT_METHOD(speakerOn) {
//  [[BearInterface getInstance] speakerOn];
}

RCT_EXPORT_METHOD(speakerOff) {
// [[BearInterface getInstance] speakerOff];
}

RCT_EXPORT_METHOD(request:(NSString *)data) {
//  [[BearInterface getInstance] request: data];
}

- (void)onJoinChatroomRS : (int32_t) result : (NSString *) chatroomID : (NSString *) userID {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.bridge.eventDispatcher sendAppEventWithName:EVENT_ON_JOIN_ROOM
                                                 body:@{@"result": [NSNumber numberWithInteger:(result)],
                                                        @"chatroomID": chatroomID,
                                                        @"userID": userID}];
  });
}

- (void)onExitChatroomRS : (int32_t) result : (NSString *) chatroomID : (NSString *) userID {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.bridge.eventDispatcher sendAppEventWithName:EVENT_ON_EXIT_ROOM
                                                 body:@{@"result": [NSNumber numberWithInteger:(result)],
                                                        @"chatroomID": chatroomID,
                                                        @"userID": userID}];
  });
}

- (void)onAudioServerConnect {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.bridge.eventDispatcher sendAppEventWithName:EVENT_ON_CONNECT
                                                 body:@true];
  });
}

- (void)onAudioServerDisconnect {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.bridge.eventDispatcher sendAppEventWithName:EVENT_ON_DISCONNECT
                                                 body:@true];
  });
}

- (void)onAudioServerBroadcast : (NSString *) resultJson {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.bridge.eventDispatcher sendAppEventWithName:EVENT_ON_BROADCAST
                                                 body:resultJson];
  });
}

- (void)onSetMessageToServerRS : (int32_t) result : (int32_t) message {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.bridge.eventDispatcher sendAppEventWithName:EVENT_ON_SEND_MESSAGE
                                                 body:@{@"result": [NSNumber numberWithInteger:(result)],
                                                        @"message": [NSNumber numberWithInteger:(message)]}];
  });
}

- (void)onMatchingChatroomRS : (int32_t) result : (NSString *) chatroomID : (NSString *) userID {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.bridge.eventDispatcher sendAppEventWithName:EVENT_ON_MATCHING_ROOM
                                                 body:@{@"result": [NSNumber numberWithInteger:(result)],
                                                        @"chatroomID": chatroomID,
                                                        @"userID": userID}];
  });
}

- (void)onSetGradeToServerRS : (int32_t) result {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.bridge.eventDispatcher sendAppEventWithName:EVENT_ON_SET_GRADE
                                                 body:@{@"result": [NSNumber numberWithInteger:(result)]}];
  });
}

- (void)onRequestRS : (int32_t) result : (NSString *) resultJson {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self.bridge.eventDispatcher sendAppEventWithName:EVENT_ON_REQUEST
                                                 body:@{@"result": [NSNumber numberWithInteger:(result)],
                                                        @"data": resultJson}];
  });
}

@end
